src/hotspot/share/opto/predicates.cpp

```diff
@@ -41,31 +41,40 @@ Node* AssertionPredicates::find_entry(Node* start_proj) {
   return entry;
 }
 
-// An Assertion Predicate has always a true projection on the success path.
-bool may_be_assertion_predicate_if(const Node* node) {
+// An Assertion Predicate has always a true projection on the success path and a Halt node on the failure path. Usually,
+// Assertion Predicates have their dedicated opaque nodes as bool input, except when an Assertion Predicate was killed
+// as part of loop splitting or during IGVN. In this case, it has a ConI input which we also treat as a valid Assertion
+// Predicate. Note that during IGVN, we could also have a dying Assertion Predicate with only one If projection left.
+// We will treat such a predicate as Runtime Predicate to skip over it during predicate iteration and not process it
+// further as Assertion Predicate.
+bool AssertionPredicate::may_be_assertion_predicate_if(const Node* node) {
   assert(node != nullptr, "should not be null");
-  return node->is_IfTrue() && RegularPredicate::may_be_predicate_if(node->as_IfProj());
+  return node->is_IfTrue() &&
+         RegularPredicate::may_be_predicate_if(node->as_IfProj()) &&
+         node->in(0)->outcnt() == 2 &&
+         has_halt(node);
 }
 
 bool AssertionPredicate::is_predicate(const Node* maybe_success_proj) {
   if (!may_be_assertion_predicate_if(maybe_success_proj)) {
     return false;
   }
-  return has_assertion_predicate_opaque(maybe_success_proj) && has_halt(maybe_success_proj);
+  return has_assertion_predicate_opaque_or_con_input(maybe_success_proj);
```

Comment:
Quick control question: Could skipping over constants also mean that we traverse up too far at some point? Like skipping not not just the predicates but also an unrelated check that is about to constant fold? I suppose that should not really create issues?